Add support for subject alternative names when requesting certificates

ean opened the pull request.

Commit:
Add support for subject alternative names when requesting certificates

luna-duclos commented:
This will need to wait until xenolf/lego#356 is merged

whereisaaron commented:
Looks great @ean 👏 

Thanks @luna-duclos I was going to note the same thing. I saw ACMESharp users were also having trouble with SAN certificates and this rate limit. Hopefully xenolf/lego#356 will fix that.

> The “new-reg”, “new-authz” and “new-cert” endpoints have an Overall Requests limit of 20 per second. All other endpoints have an Overall Requests limit of 2000 per second.

There is this newly introduced failures rate limit to watch out for, if we retry failed validations too quickly. But that's nothing specific to this PR.

> We will soon (February 2017) introduce a Failed Validation limit of 5 failures per account, per hostname, per hour. This limit will be higher on staging so you can use staging to debug connectivity problems. 

https://letsencrypt.org/docs/rate-limits/

luna-duclos commented:
xenolf/lego's master should now have a fix for the rate limit.

luna-duclos commented:
@ean: Are there any further changes you would like to make to this before I review/merge it ?

ean commented:
@luna-duclos please review, I have been running this at work for about a month now, and it seems stable.

rosskukulinski commented:
I've been running @ean's code for SANs as well as a patched xenolf/lego for a few weeks now and haven't had any problems.  Thank you @ean!

euank commented on processor.go:
```diff
@@ -440,14 +500,21 @@ func (p *CertProcessor) processCertificate(cert Certificate) (processed bool, er
-	if certDetailsRaw == nil || acmeCert.DomainName == "" {
+	domains := make([]string, 1+len(altNames))
+	domains[0] = cert.Spec.Domain
+	for i, altName := range altNames {
```
This for loop could be `copy(domains[1:], altNames)` instead; slightly more readable I think.

euank replied:
Actually, after thinking a second more, I think the most readable option is:

`domains := append([]string{cert.Spec.Domain}, altNames...)`

euank commented on processor.go:
```diff
@@ -649,22 +723,13 @@ func (p *CertProcessor) processIngress(ingress v1beta1.Ingress) {
-				Type:           "Warning",
-			})
-			continue
+		altNames := make([]string, len(tls.Hosts)-1)
```
It's probably okay to just slice `tls.Hosts`; `altNames = tls.Hosts[1:]` The tls object should be considered immutable anyways (and indeed processCertificate does copy it).

euank commented on processor.go:
```diff
@@ -331,6 +332,57 @@ func (p *CertProcessor) secretName(cert Certificate) string {
+		return nil, errors.Wrapf(err, "Error while fetching altnames from database for domain %v", cert.Spec.Domain)
+	}
+	if altNamesRaw == nil {
+		altNamesRaw, err = json.Marshal([]string{})
```
it seems like it would be easier to `return []string{}, nil` here rather than go to the trouble of marshal+unmarshal to do the same.

euank reviewed (commented):
I added a few style nits, but this overall looks good I think.

The one other thing to consider would be updating the glide.lock file (and possibly checking in the vendor tree) since apparently there are upstream lego fixes that would help.

Commit:
Simplify array handling

Commit:
Early return when returning default value

ean commented:
Thanks @euank, updated the code to match your comments. It ended up being a lot easier on the eyes.

euank commented:
This PR also needs to remove the length check here: https://github.com/PalmStoneGames/kube-cert-manager/blob/bc5064a21b1c9ec0c21590ce41b47526884ca647/processor.go#L619-L621

Without that change, I have it erroneously delete the secret for an ingress the next sync-interval after creating it.

Maybe it should just be flipped to a `< 1` check to ensure the bit below doesn't panic.

Commit:
Support altnames for ingress certificates

ean commented:
@euank - Thanks for testing this. I updated the ingress certificate logic now. I also saw that AltNames were not added to the certificate spec, which I also fixed. Could you give it another test? I don't have any ingress setup to test this on.

euank commented:
I'm not seeing any other issues, though I'm running it rebased on top of my client-go change as well.

euank commented:
Ping @luna-duclos, what's the next step for this?

luna-duclos commented:
I've been away for easter! I will be merging those, and the other open PRs, pretty soon, then we can have a test period, and then when they prove stable, make a new release and move the repo.

luna-duclos commented on processor.go:
```diff
@@ -331,6 +332,54 @@ func (p *CertProcessor) secretName(cert Certificate) string {
+		return nil, errors.Wrapf(err, "Error while fetching altnames from database for domain %v", cert.Spec.Domain)
+	}
+	if altNamesRaw == nil {
+		return []string{}, nil
```
Can just return nil here instead of an empty initialized slice

ean replied:
fixed

luna-duclos commented on processor.go:
```diff
@@ -350,8 +399,16 @@ func (p *CertProcessor) processCertificate(cert Certificate) (processed bool, er
+	altNames := normalizeHostnames(cert.Spec.AltNames)
+	storedAltNames, err := p.getStoredAltNames(cert)
+	if err != nil {
+		return false, err
```
This error should be wrapped

ean replied:
fixed

luna-duclos reviewed (approved):
Two small comments, if you don't mind looking at them, I don't mind fixing them myself either if you lack the time, let me know! :)

Commit:
Wrap error and return nil as empty slice
